gcc/rust/resolve/rust-ast-resolve.cc

```diff
@@ -0,0 +1,252 @@
+// Copyright (C) 2020 Free Software Foundation, Inc.
+
+// This file is part of GCC.
+
+// GCC is free software; you can redistribute it and/or modify it under
+// the terms of the GNU General Public License as published by the Free
+// Software Foundation; either version 3, or (at your option) any later
+// version.
+
+// GCC is distributed in the hope that it will be useful, but WITHOUT ANY
+// WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+// for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with GCC; see the file COPYING3.  If not see
+// <http://www.gnu.org/licenses/>.
+
+#include "rust-ast-resolve.h"
+#include "rust-ast-resolve-toplevel.h"
+#include "rust-ast-resolve-item.h"
+#include "rust-ast-full.h"
+#include "rust-tyty.h"
+
+#define MKBUILTIN_TYPE(_X, _R, _TY)                                            \
```

Comment:
This is more of a stylistic choice than anything else, but making this a template member function (if required - otherwise just a regular member function) might be better than having it as a macro. I believe that doing so would be possible as long as `_R` is passed in by reference - no other modification of parameters seems to occur.